lib/honeybadger/events_worker.ex

```diff
@@ -0,0 +1,241 @@
+    timer_ref = schedule_flush(state)
+    {:ok, %{state | timer_ref: timer_ref, last_dropped_log: System.monotonic_time(:millisecond)}}
+  end
+
+  @impl true
+  def handle_call({:state}, _from, %State{} = state) do
+    {:reply, state, state}
+  end
+
+  @impl true
+  def handle_cast({:push, event}, %State{} = state) do
+    if total_event_count(state) >= state.max_queue_size do
+      {:noreply, %{state | dropped_events: state.dropped_events + 1}}
+    else
+      queue = [event | state.queue]
+
+      if length(queue) >= state.batch_size do
+        batches = :queue.in(%{batch: Enum.reverse(queue), attempts: 0}, state.batches)
+        state = attempt_send(%{state | queue: [], batches: batches})
+        {:noreply, state}
+      else
+        {:noreply, %{state | queue: queue}}
+      end
+    end
+  end
+
+  @impl true
+  # With empty batches and queue just reschedule another flush
+  def handle_info(:flush, %State{batches: [], queue: []} = state) do
+    {:noreply, %{state | timer_ref: schedule_flush(state)}}
+  end
+
+  @impl true
+  # With an empty queue, just attempt to send the batches
+  def handle_info(:flush, %State{queue: []} = state) do
+    {:noreply, attempt_send(state)}
+  end
+
+  @impl true
+  # Flush the queue into a new batch and attempt to send it
+  def handle_info(:flush, %State{queue: queue} = state) do
+    batches = :queue.in(%{batch: Enum.reverse(queue), attempts: 0}, state.batches)
+    {:noreply, attempt_send(%{state | queue: [], batches: batches})}
+  end
+
+  @impl true
+  def terminate(_reason, %State{timer_ref: timer_ref} = state) do
+    if timer_ref, do: Process.cancel_timer(timer_ref)
+
+    Logger.debug("[Honeybadger] Terminating with #{total_event_count(state)} events unsent")
+
+    # Best effort final flush (inline)
+    final_state =
+      if state.queue != [] do
+        batch = %{batch: Enum.reverse(state.queue), attempts: 0}
+        %{state | queue: [], batches: :queue.in(batch, state.batches)}
+      else
+        state
+      end
+
+    _ = attempt_send(final_state)
+
+    :ok
+  end
+
+  @spec schedule_flush(State.t()) :: reference() | nil
+  # Schedules a flush timer (or reuses the current one when throttled with no
+  # new events).
+  defp schedule_flush(%State{timer_ref: timer_ref, throttling: true}), do: timer_ref
+
+  defp schedule_flush(%State{timer_ref: timer_ref, timeout: timeout}) do
+    if timer_ref, do: Process.cancel_timer(timer_ref)
+    Process.send_after(self(), :flush, timeout)
+  end
+
+  @spec attempt_send(State.t()) :: State.t()
+  # Sends pending batches, handling retries and throttling, and schedules the
+  # next flush.
+  defp attempt_send(%State{timer_ref: timer_ref} = state) do
+    if timer_ref, do: Process.cancel_timer(timer_ref)
+
+    {new_batches_list, throttling} =
+      Enum.reduce(:queue.to_list(state.batches), {[], false}, fn
+        # If already throttled, skip sending and retain the batch.
+        b, {acc, true} ->
+          {acc ++ [b], true}
```

Comment:
Can this be refactored into a `Enum.reduce_while` instead?

Reply:
I had this as a reduce_while at first, but I changed it because I didn't really have a halt branch.